pkg/storage/series/series_set.go

```diff
@@ -79,52 +83,136 @@ func (c *ConcreteSeries) Iterator(_ chunkenc.Iterator) chunkenc.Iterator {
+}
+
 func (c *concreteSeriesIterator) Seek(t int64) chunkenc.ValueType {
-	c.cur = sort.Search(len(c.series.samples), func(n int) bool {
+	oldTime, oldType := c.atType()
+	if oldTime >= t { // only advance via Seek
+		return oldType
+	}
+
+	c.curFloat = sort.Search(len(c.series.samples), func(n int) bool {
 		return c.series.samples[n].Timestamp >= model.Time(t)
 	})
-	if c.cur < len(c.series.samples) {
+	c.curHisto = sort.Search(len(c.series.histograms), func(n int) bool {
+		return c.series.histograms[n].Timestamp >= t
+	})
+
+	if c.curFloat >= len(c.series.samples) && c.curHisto >= len(c.series.histograms) {
+		return chunkenc.ValNone
+	}
+	if c.curFloat >= len(c.series.samples) {
+		c.atHisto = true
+		return c.atTypeHisto()
+	}
+	if c.curHisto >= len(c.series.histograms) {
+		c.atHisto = false
 		return chunkenc.ValFloat
 	}
-	return chunkenc.ValNone
+	if int64(c.series.samples[c.curFloat].Timestamp) < c.series.histograms[c.curHisto].Timestamp {
+		c.curHisto--
+		c.atHisto = false
+		return chunkenc.ValFloat
+	}
+	c.curFloat--
+	c.atHisto = true
+	return c.atTypeHisto()
 }
 
 func (c *concreteSeriesIterator) At() (t int64, v float64) {
-	s := c.series.samples[c.cur]
+	if c.atHisto {
+		panic(errors.New("concreteSeriesIterator: Calling At() when cursor is at histogram"))
+	}
+	s := c.series.samples[c.curFloat]
 	return int64(s.Timestamp), float64(s.Value)
 }
 
 func (c *concreteSeriesIterator) Next() chunkenc.ValueType {
-	c.cur++
-	if c.cur < len(c.series.samples) {
+	if c.curFloat+1 >= len(c.series.samples) && c.curHisto+1 >= len(c.series.histograms) {
+		c.curFloat++
+		c.curHisto++
+		return chunkenc.ValNone
+	}
+	if c.curFloat+1 >= len(c.series.samples) {
+		c.curHisto++
+		c.atHisto = true
+		return c.atTypeHisto()
+	}
+	if c.curHisto+1 >= len(c.series.histograms) {
+		c.curFloat++
+		c.atHisto = false
 		return chunkenc.ValFloat
 	}
-	return chunkenc.ValNone
+	if int64(c.series.samples[c.curFloat+1].Timestamp) < c.series.histograms[c.curHisto+1].Timestamp {
+		c.curFloat++
+		c.atHisto = false
+		return chunkenc.ValFloat
+	}
+	c.curHisto++
+	c.atHisto = true
+	return c.atTypeHisto()
 }
 
 func (c *concreteSeriesIterator) AtHistogram() (int64, *histogram.Histogram) {
-	panic(errors.New("concreteSeriesIterator: AtHistogram not implemented"))
+	if !c.atHisto {
+		panic(errors.New("concreteSeriesIterator: Calling AtHistogram() when cursor is not at histogram"))
+	}
+	h := c.series.histograms[c.curHisto]
+	return h.Timestamp, mimirpb.FromHistogramProtoToHistogram(&h)
 }
 
 func (c *concreteSeriesIterator) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
-	panic(errors.New("concreteSeriesIterator: AtHistogram not implemented"))
+	if !c.atHisto {
+		panic(errors.New("concreteSeriesIterator: Calling AtFloatHistogram() when cursor is not at histogram"))
+	}
+	h := c.series.histograms[c.curHisto]
+	// Should we automatically convert to float if the histogram happens to be integer histogram?
```

Comment:
I think yes, since this iterator is a variant of the iterators that can be used for the promql engine: https://github.com/grafana/mimir/blob/651a8edda93f1532ed8997d89d9c930b9fd4f3fc/pkg/querier/querier.go#L92